waspc/src/Wasp/SemanticVersion.hs

```diff
@@ -0,0 +1,31 @@
+module Wasp.SemanticVersion
+  ( SemanticVersion (..),
+    VersionBounds (..),
+    isVersionInBounds,
+  )
+where
+
+import Text.Printf (printf)
+
+data SemanticVersion = SemanticVersion
+  { major :: Int,
```

Comment:
Here you say these are `Int`. But, we know they will be non-negative, right?
Sooo, you could use more exact type, https://hackage.haskell.org/package/base-4.10.0.0/docs/Numeric-Natural.html :D. I have to admit I never used it, but it is in base and it looks to fit the role! So maybe it is a fun opportunity to give it a try. Cool thing is you will know you don't have a negative number in here.

Reply:
And I was thinking the exact opposite (i.e., loosening the type). Version names can sometimes be strange, for example, `v1.4.0-beta2` or `v1.0.0-rc`.

I thought of making these things strings, but that natural number type does look interesting, I think we can have that for now, why not.

Reply:
I am ok with whatever you choose for now, I thought Natural numbers might be interesting :D, but you are right about the versions being all types of things!
Although I think stuff mostly gets added at the end with `-<some_string>`, so in theory the type could then be

```hs
  major :: Natural,
  minor :: Natural,
  patch :: Natural,
  suffix :: String
```

Reply:
I'd rather stay away from the suffix for now. Adding it would mean having to ensure the right compatibility rules in the `isVersionInBounds` function (and they get even more complicated with the suffix).

Is it OK to add the suffix if we ever end up needing such precision (after all, this is used only for our packages)?

Reply:
Oh yes exactly, I wouldn't add it now! I just wanted to say that even if we have such strange packages, it might still have sense to have those three Natural fields, and then we can just add one String field.
So yes I would certainly not add it for now!